examples/gno.land/r/demo/profile/profile.gno

```diff
@@ -0,0 +1,428 @@
+package profile
+
+import (
+	"bytes"
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/mux"
+	"gno.land/p/demo/ufmt"
+	"std"
+	"time"
+)
+
+type WebsiteInfo struct {
+	Name string
+	URL  string
+}
+
+type UserProfile struct {
+	Username            string
+	Address             std.Address
+	AvatarURL           string
+	Age                 string
+	Gender              string
+	Websites            *avl.Tree // Name -> URL
+	Country             string
+	City                string
+	ModificationHistory map[string][]string
+}
+
+var profiles *avl.Tree      // address -> UserProfile
+var usernameIndex *avl.Tree // username -> address
+var router *mux.Router
+
+func init() {
+	profiles = avl.NewTree()
+	usernameIndex = avl.NewTree()
+	router = mux.NewRouter()
+
+	router.HandleFunc("", homeHandler)
+	router.HandleFunc("info/{username}", profileHandler)
+	router.HandleFunc("history/{username}", historyHandler)
+}
+
+func CreateProfile(username, avatarurl, age, gender, websiteName, websiteURL, country, city string) {
+	caller := std.GetOrigCaller()
+
+	// Check if a profile already exists for this address
+	if _, exists := profiles.Get(caller.String()); exists {
+		panic("a profile already exists for this address")
+	}
+
+	// Check if the username is already taken
+	if username != "" {
+		_, exists := usernameIndex.Get(username)
+		if exists {
+			panic("Username is already taken")
+		}
+	}
+
+	websites := avl.NewTree()
+	websites.Set(websiteName, websiteURL)
+
+	profile := UserProfile{
```

Comment:
I imagined `r/profile` to not have a struct with predefined fields like this, but instead to have a way to add arbitrary fields to an address in a similar fashion to `context.WithValue` (https://pkg.go.dev/context#example-WithValue).

This package should focus on creating this system and adding helpers to manage the "most common" fields. These helpers should define unexported keys and constructors that will demonstrate how to extend profiles with more structured fields by any other packages.

Reply:
I've refactored my code so that the creation of profile basic can only be done with a username linked to an address. I've added an avl.tree to the UserProfile structure and an AddFields function that lets you add one or more other fields with a key and one or more values, allowing the field to be extended via another realm or package. I've still got a few things to modify and improve, but I wanted to get some advice on how I go about adding extra fields.

- https://github.com/gnolang/gno/pull/1983/commits/19d5b2364ba3202227871a7bd95716cc78f59339